File: cpp-client/deephaven/tests/src/select_test.cc
/*
 * Copyright (c) 2016-2025 Deephaven Data Labs and Patent Pending
 */
#include <iostream>
#include "deephaven/third_party/catch.hpp"
#include "deephaven/tests/test_util.h"
#include "deephaven/dhcore/types.h"
#include "deephaven/dhcore/utility/utility.h"
#include "deephaven/third_party/fmt/format.h"

using deephaven::client::Client;
using deephaven::client::TableHandle;
using deephaven::client::utility::TableMaker;
using deephaven::dhcore::DateTime;
using deephaven::dhcore::LocalDate;
using deephaven::dhcore::LocalTime;
using deephaven::dhcore::DeephavenConstants;

namespace deephaven::client::tests {
TEST_CASE("Support all types", "[select]") {
  auto tm = TableMakerForTests::Create();

  std::vector<bool> bool_data;
  std::vector<char16_t> char_data;
  std::vector<int8_t> byte_data;
  std::vector<int16_t> short_data;
  std::vector<int32_t> int_data;
  std::vector<int64_t> long_data;
  std::vector<float> float_data;
  std::vector<double> double_data;
  std::vector<std::string> string_data;
  std::vector<DateTime> date_time_data;
  std::vector<LocalDate> local_date_data;
  std::vector<LocalTime> local_time_data;

  const int start_value = -8;
  const int end_value = 8;
  for (auto i = start_value; i != end_value; ++i) {
    bool_data.push_back((i % 2) == 0);
    char_data.push_back(i * 10);
    byte_data.push_back(i * 11);
    short_data.push_back(i * 1000);
    int_data.push_back(i * 1'000'000);
    long_data.push_back(static_cast<int64_t>(i) * 1'000'000'000);
    float_data.push_back(i * 123.456F);
    double_data.push_back(i * 987654.321);
    string_data.push_back(fmt::format("test {}", i));
    date_time_data.push_back(DateTime::FromNanos(i));
    local_date_data.push_back(LocalDate::FromMillis(i * 86400 * 1000));
    local_time_data.push_back(LocalTime::FromNanos(1000 + i));  // nanos argument cannot be negative
  }

  TableMaker maker;
  maker.AddColumn("boolData", bool_data);
  maker.AddColumn("charData", char_data);
  maker.AddColumn("byteData", byte_data);
  maker.AddColumn("shortData", short_data);
  maker.AddColumn("intData", int_data);
  maker.AddColumn("longData", long_data);
  maker.AddColumn("floatData", float_data);
  maker.AddColumn("doubleData", double_data);
  maker.AddColumn("stringData", string_data);
  maker.AddColumn("dateTimeData", date_time_data);
  maker.AddColumn("localDateData", local_date_data);
  maker.AddColumn("localTimeData", local_time_data);

  auto t = maker.MakeTable(tm.Client().GetManager());

  std::cout << t.Stream(true) << '\n';

  TableComparerForTests::Compare(maker, t);
}

TEST_CASE("Create / Update / fetch a Table", "[select]") {
  auto tm = TableMakerForTests::Create();

  TableMaker maker;
  maker.AddColumn<int32_t>("IntValue", {0, 1, 2, 3, 4, 5, 6, 7, 8, 9});
  maker.AddColumn<double>("DoubleValue", {0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9});
  maker.AddColumn<std::string>("StringValue", {"zero", "one", "two", "three", "four", "five", "six", "seven",
      "eight", "nine"});
  auto t = maker.MakeTable(tm.Client().GetManager());
  auto t2 = t.Update("Q2 = IntValue * 100");
  auto t3 = t2.Update("Q3 = Q2 + 10");
  auto t4 = t3.Update("Q4 = Q2 + 100");

  // Reuse the already-populated maker; keep the existing columns and add a few more.
  maker.AddColumn<int32_t>("Q2", {0, 100, 200, 300, 400, 500, 600, 700, 800, 900});
  maker.AddColumn<int32_t>("Q3", {10, 110, 210, 310, 410, 510, 610, 710, 810, 910});
  maker.AddColumn<int32_t>("Q4", {100, 200, 300, 400, 500, 600, 700, 800, 900, 1000});
  TableComparerForTests::Compare(maker, t4);
}


TEST_CASE("Select a few columns", "[select]") {
  auto tm = TableMakerForTests::Create();
  auto table = tm.Table();

  auto t1 = table.Where("ImportDate == `2017-11-01` && Ticker == `AAPL`")
      .Select("Ticker", "Close", "Volume")
      .Head(2);

  TableMaker expected;
  expected.AddColumn<std::string>("Ticker", {"AAPL", "AAPL"});
  expected.AddColumn<double>("Close", {23.5, 24.2});
  expected.AddColumn<int64_t>("Volume", {100000, 250000});

  TableComparerForTests::Compare(expected, t1);
}

TEST_CASE("LastBy + Select", "[select]") {
  auto tm = TableMakerForTests::Create();
  auto table = tm.Table();

  auto t1 = table.Where("ImportDate == `2017-11-01` && Ticker == `AAPL`")
      .LastBy("Ticker")
      .Select("Ticker", "Close", "Volume");
  std::cout << t1.Stream(true) << '\n';

  TableMaker expected;
  expected.AddColumn<std::string>("Ticker", {"AAPL"});
  expected.AddColumn<double>("Close", {26.7});
  expected.AddColumn<int64_t>("Volume", {19000});
  TableComparerForTests::Compare(expected, t1);
}

TEST_CASE("New columns", "[select]") {
  auto tm = TableMakerForTests::Create();
  auto table = tm.Table();

  // A formula expression
  auto t1 = table.Where("ImportDate == `2017-11-01` && Ticker == `AAPL`")
      .Select("MV1 = Volume * Close", "V_plus_12 = Volume + 12");

  TableMaker expected;
  expected.AddColumn<double>("MV1", {2350000, 6050000, 507300});
  expected.AddColumn<int64_t>("V_plus_12", {100012, 250012, 19012});
  TableComparerForTests::Compare(expected, t1);
}

TEST_CASE("Drop columns", "[select]") {
  auto tm = TableMakerForTests::Create();
  auto table = tm.Table();

  auto t1 = table.DropColumns({"ImportDate", "Open", "Close"});
  CHECK(2 == t1.Schema()->NumCols());
}

TEST_CASE("Simple Where", "[select]") {
  auto tm = TableMakerForTests::Create();
  auto table = tm.Table();
  auto updated = table.Update("QQQ = i");

  auto t1 = updated.Where("ImportDate == `2017-11-01` && Ticker == `IBM`")
      .Select("Ticker", "Volume");

  TableMaker expected;
  expected.AddColumn<std::string>("Ticker", {"IBM"});
  expected.AddColumn<int64_t>("Volume", {138000});
  TableComparerForTests::Compare(expected, t1);
}

TEST_CASE("Formula in the Where clause", "[select]") {
  auto tm = TableMakerForTests::Create();
  auto table = tm.Table();

  auto t1 = table.Where(
          "ImportDate == `2017-11-01` && Ticker == `AAPL` && Volume % 10 == Volume % 100")
      .Select("Ticker", "Volume");
  std::cout << t1.Stream(true) << '\n';

  TableMaker expected;
  expected.AddColumn<std::string>("Ticker", {"AAPL", "AAPL", "AAPL"});
  expected.AddColumn<int64_t>("Volume", {100000, 250000, 19000});
  TableComparerForTests::Compare(expected, t1);
}

TEST_CASE("Simple 'Where' with syntax error", "[select]") {
  auto tm = TableMakerForTests::Create();
  auto table = tm.Table();

  try {
    // String literal
    auto t1 = table.Where(")))))");
    std::cout << t1.Stream(true) << '\n';
  } catch (const std::exception &e) {
    // Expected
    fmt::print(std::cerr, "Caught *expected* exception {}\n", e.what());
    return;
  }
  throw std::runtime_error("Expected a failure, but didn't experience one");
}

TEST_CASE("WhereIn", "[select]") {
  auto tm = TableMakerForTests::Create();

  std::vector<std::string> letter_data = {"A", "C", "F", "B", "E", "D", "A"};
  std::vector<std::optional<int32_t>> number_data = { {}, 2, 1, {}, 4, 5, 3};
  std::vector<std::string> color_data = {"red", "blue", "orange", "purple", "yellow", "pink", "blue"};
  std::vector<std::optional<int32_t>> code_data = { 12, 13, 11, {}, 16, 14, {}};
  TableMaker source_maker;
  source_maker.AddColumn("Letter", letter_data);
  source_maker.AddColumn("Number", number_data);
  source_maker.AddColumn("Color", color_data);
  source_maker.AddColumn("Code", code_data);
  auto source = source_maker.MakeTable(tm.Client().GetManager());

  std::vector<std::string> filter_color_data = {"blue", "red", "purple", "white"};
  TableMaker filter_maker;
  filter_maker.AddColumn("Colors", filter_color_data);
  auto filter = filter_maker.MakeTable(tm.Client().GetManager());

  auto result = source.WhereIn(filter, {"Color = Colors"});

  TableMaker expected;
  expected.AddColumn<std::string>("Letter", {"A", "C", "B", "A"});
  expected.AddColumn<std::optional<int32_t>>("Number", { {}, 2, {}, 3});
  expected.AddColumn<std::string>("Color", {"red", "blue", "purple", "blue"});
  expected.AddColumn<std::optional<int32_t>>("Code", { 12, 13, {}, {}});
  TableComparerForTests::Compare(expected, result);
}

TEST_CASE("LazyUpdate", "[select]") {
  auto tm = TableMakerForTests::Create();

  TableMaker maker;
  maker.AddColumn<std::string>("A", {"The", "At", "Is", "On"});
  maker.AddColumn<int32_t>("B", {1, 2, 3, 4});
  maker.AddColumn<int32_t>("C", {5, 2, 5, 5});
  auto source = maker.MakeTable(tm.Client().GetManager());

  auto result = source.LazyUpdate({"Y = sqrt(C)"});

  // Reuse maker and add one more column.
  maker.AddColumn<double>("Y", {std::sqrt(5), std::sqrt(2), std::sqrt(5), std::sqrt(5)});
  TableComparerForTests::Compare(maker, result);
}

TEST_CASE("SelectDistinct", "[select]") {
  auto tm = TableMakerForTests::Create();

  std::vector<std::string> a_data = {"apple", "apple", "orange", "orange", "plum", "grape"};
  std::vector<int32_t> b_data = {1, 1, 2, 2, 3, 3};
  TableMaker source_maker;
  source_maker.AddColumn("A", a_data);
  source_maker.AddColumn("B", b_data);
  auto source = source_maker.MakeTable(tm.Client().GetManager());

  auto result = source.SelectDistinct({"A"});

  std::cout << result.Stream(true) << '\n';

  std::vector<std::string> expected_data = {"apple", "orange", "plum", "grape"};
  TableMaker expected;
  expected.AddColumn<std::string>("A", {"apple", "orange", "plum", "grape"});
  TableComparerForTests::Compare(expected, result);
}
}  // namespace deephaven::client::tests

File: csharp/client/Dh_NetClientTests/SelectTest.cs
//
// Copyright (c) 2016-2025 Deephaven Data Labs and Patent Pending
//
using Deephaven.Dh_NetClient;

namespace Deephaven.Dh_NetClientTests;

public class SelectTest {
  [Test]
  public async Task TestSupportAllTypes() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());

    var boolData = new List<bool>();
    var charData = new List<char>();
    var byteData = new List<sbyte>();
    var shortData = new List<short>();
    var intData = new List<int>();
    var longData = new List<long>();
    var floatData = new List<float>();
    var doubleData = new List<double>();
    var stringData = new List<string>();
    var dateTimeData = new List<DateTimeOffset>();
    var dateOnlyData = new List<DateOnly>();
    var timeOnlyData = new List<TimeOnly>();

    const int startValue = -8;
    const int endValue = 8;
    for (var i = startValue; i != endValue; ++i) {
      boolData.Add((i % 2) == 0);
      charData.Add((char)(i * 10));
      byteData.Add((sbyte)(i * 11));
      shortData.Add((short)(i * 1000));
      intData.Add(i * 1_000_000);
      longData.Add(i * (long)1_000_000_000);
      floatData.Add(i * 123.456F);
      doubleData.Add(i * 987654.321);
      stringData.Add($"test {i}");
      dateTimeData.Add(DateTimeOffset.FromUnixTimeMilliseconds(i));
      dateOnlyData.Add(DateOnly.FromDayNumber(i + 1000));
      timeOnlyData.Add(TimeOnly.FromTimeSpan(TimeSpan.FromMilliseconds(i + 1000)));
    }

    var tm = new TableMaker();
    tm.AddColumn("boolData", boolData);
    tm.AddColumn("charData", charData);
    tm.AddColumn("byteData", byteData);
    tm.AddColumn("shortData", shortData);
    tm.AddColumn("intData", intData);
    tm.AddColumn("longData", longData);
    tm.AddColumn("floatData", floatData);
    tm.AddColumn("doubleData", doubleData);
    tm.AddColumn("stringData", stringData);
    tm.AddColumn("dateTimeData", dateTimeData);
    tm.AddColumn("dateOnlyData", dateOnlyData);
    tm.AddColumn("timeOnlyData", timeOnlyData);

    var th = tm.MakeTable(ctx.Client.Manager);
    Console.WriteLine(th.ToString(true));

    await Assert.That(() => TableComparer.AssertSame(tm, th)).ThrowsNothing();
  }

  [Test]
  public async Task TestCreateUpdateFetchATable() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());

    var tm = new TableMaker();
    tm.AddColumn("IntValue", [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
    tm.AddColumn("DoubleValue", [0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9]);
    tm.AddColumn("StringValue", [
      "zero", "one", "two", "three", "four", "five", "six", "seven",
      "eight", "nine"
    ]);
    var t = tm.MakeTable(ctx.Client.Manager);
    var t2 = t.Update("Q2 = IntValue * 100");
    var t3 = t2.Update("Q3 = Q2 + 10");
    var t4 = t3.Update("Q4 = Q2 + 100");

    tm.AddColumn("Q2", [0, 100, 200, 300, 400, 500, 600, 700, 800, 900]);
    tm.AddColumn("Q3", [10, 110, 210, 310, 410, 510, 610, 710, 810, 910]);
    tm.AddColumn("Q4", [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000]);
    await Assert.That(() => TableComparer.AssertSame(tm, t4)).ThrowsNothing();
  }

  [Test]
  public async Task TestSelectAFewColumns() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    var t1 = table.Where("ImportDate == `2017-11-01` && Ticker == `AAPL`")
      .Select("Ticker", "Close", "Volume")
      .Head(2);

    var tm = new TableMaker();
    tm.AddColumn("Ticker", ["AAPL", "AAPL"]);
    tm.AddColumn("Close", [23.5, 24.2]);
    tm.AddColumn("Volume", [(Int64)100000, 250000]);
    await Assert.That(() => TableComparer.AssertSame(tm, t1)).ThrowsNothing();
  }

  [Test]
  public async Task TestLastByAndSelect() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    var t1 = table.Where("ImportDate == `2017-11-01` && Ticker == `AAPL`")
      .LastBy("Ticker")
      .Select("Ticker", "Close", "Volume");
    Console.WriteLine(t1.ToString(true));

    var tm = new TableMaker();
    tm.AddColumn("Ticker", ["AAPL"]);
    tm.AddColumn("Close", [26.7]);
    tm.AddColumn("Volume", [(Int64)19000]);
    await Assert.That(() => TableComparer.AssertSame(tm, t1)).ThrowsNothing();
  }

  [Test]
  public async Task TestNewColumns() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    // A formula expression
    var t1 = table.Where("ImportDate == `2017-11-01` && Ticker == `AAPL`")
      .Select("MV1 = Volume * Close", "V_plus_12 = Volume + 12");

    var tm = new TableMaker();
    tm.AddColumn("MV1", [(double)2350000, 6050000, 507300]);
    tm.AddColumn("V_plus_12", [(Int64)100012, 250012, 19012]);
    await Assert.That(() => TableComparer.AssertSame(tm, t1)).ThrowsNothing();
  }

  [Test]
  public async Task TestDropColumns() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    var t1 = table.DropColumns("ImportDate", "Open", "Close");
    await Assert.That(table.Schema.FieldsList.Count).IsEqualTo(5);
    await Assert.That(t1.Schema.FieldsList.Count).IsEqualTo(2);
  }

  [Test]
  public async Task TestSimpleWhere() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;
    var updated = table.Update("QQQ = i");

    var t1 = updated.Where("ImportDate == `2017-11-01` && Ticker == `IBM`")
      .Select("Ticker", "Volume");

    var tm = new TableMaker();
    tm.AddColumn("Ticker", ["IBM"]);
    tm.AddColumn("Volume", [(Int64)138000]);
    await Assert.That(() => TableComparer.AssertSame(tm, t1)).ThrowsNothing();
  }

  [Test]
  public async Task TestFormulaInTheWhereClause() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    var t1 = table.Where(
        "ImportDate == `2017-11-01` && Ticker == `AAPL` && Volume % 10 == Volume % 100")
      .Select("Ticker", "Volume");
    Console.WriteLine(t1.ToString(true));

    var tm = new TableMaker();
    tm.AddColumn("Ticker", ["AAPL", "AAPL", "AAPL"]);
    tm.AddColumn("Volume", [(Int64)100000, 250000, 19000]);
    await Assert.That(() => TableComparer.AssertSame(tm, t1)).ThrowsNothing();
  }

  [Test]
  public async Task TestSimpleWhereWithSyntaxError() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    await Assert.That(() => {
      var t1 = table.Where(")))))");
      Console.WriteLine(t1.ToString(true));
    }).Throws<AggregateException>();
  }

  [Test]
  public async Task TestWhereIn() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());

    var sourceMaker = new TableMaker();
    sourceMaker.AddColumn("Letter", ["A", "C", "F", "B", "E", "D", "A"]);
    sourceMaker.AddColumn("Number", [(Int32?)null, 2, 1, null, 4, 5, 3]);
    sourceMaker.AddColumn("Color", ["red", "blue", "orange", "purple", "yellow", "pink", "blue"]);
    sourceMaker.AddColumn("Code", [(Int32?)12, 13, 11, null, 16, 14, null]);

    var source = sourceMaker.MakeTable(ctx.Client.Manager);

    var filterMaker = new TableMaker();
    filterMaker.AddColumn("Colors", ["blue", "red", "purple", "white"]);
    var filter = filterMaker.MakeTable(ctx.Client.Manager);

    var result = source.WhereIn(filter, "Color = Colors");

    var expected = new TableMaker();
    expected.AddColumn("Letter", ["A", "C", "B", "A"]);
    expected.AddColumn("Number", [(Int32?)null, 2, null, 3]);
    expected.AddColumn("Color", ["red", "blue", "purple", "blue"]);
    expected.AddColumn("Code", [(Int32?)12, 13, null, null]);
    await Assert.That(() => TableComparer.AssertSame(expected, result)).ThrowsNothing();
  }

  [Test]
  public async Task TestLazyUpdate() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());

    var sourceMaker = new TableMaker();
    sourceMaker.AddColumn("A", ["The", "At", "Is", "On"]);
    sourceMaker.AddColumn("B", [1, 2, 3, 4]);
    sourceMaker.AddColumn("C", [5, 2, 5, 5]);
    var source = sourceMaker.MakeTable(ctx.Client.Manager);

    var result = source.LazyUpdate("Y = sqrt(C)");

    var expected = new TableMaker();
    expected.AddColumn("A", ["The", "At", "Is", "On"]);
    expected.AddColumn("B", [1, 2, 3, 4]);
    expected.AddColumn("C", [5, 2, 5, 5]);
    expected.AddColumn("Y", [Math.Sqrt(5), Math.Sqrt(2), Math.Sqrt(5), Math.Sqrt(5)]);
    await Assert.That(() => TableComparer.AssertSame(expected, result)).ThrowsNothing();
  }

  [Test]
  public async Task TestSelectDistinct() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());

    var sourceMaker = new TableMaker();
    sourceMaker.AddColumn("A", ["apple", "apple", "orange", "orange", "plum", "grape"]);
    sourceMaker.AddColumn("B", [1, 1, 2, 2, 3, 3]);
    var source = sourceMaker.MakeTable(ctx.Client.Manager);

    var result = source.SelectDistinct("A");
    Console.WriteLine(result.ToString(true));

    var expected = new TableMaker();
    expected.AddColumn("A", ["apple", "orange", "plum", "grape"]);
    await Assert.That(() => TableComparer.AssertSame(expected, result)).ThrowsNothing();
  }
}
